src/source/Crypto/Tls_mbedtls.c

```diff
@@ -0,0 +1,241 @@
+
+    /* init and send handshake */
+    tlsSessionChangeState(pTlsSession, TLS_SESSION_STATE_CONNECTING);
+    sslRet = mbedtls_ssl_handshake(&pTlsSession->sslCtx);
+    CHK(sslRet == MBEDTLS_ERR_SSL_WANT_READ || sslRet == MBEDTLS_ERR_SSL_WANT_WRITE, STATUS_SSL_CTX_CREATION_FAILED);
+    LOG_MBEDTLS_ERROR("mbedtls_ssl_handshake", sslRet);
+
+CleanUp:
+
+    CHK_LOG_ERR(retStatus);
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS tlsSessionProcessPacket(PTlsSession pTlsSession, PBYTE pData, UINT32 bufferLen, PUINT32 pDataLen)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    INT32 sslRet, readBytes = 0;
+    BOOL iterate = TRUE;
+    PIOBuffer pReadBuffer;
+
+    CHK(pTlsSession != NULL && pData != NULL && pDataLen != NULL, STATUS_NULL_ARG);
+    CHK(pTlsSession->state != TLS_SESSION_STATE_NEW, STATUS_SOCKET_CONNECTION_NOT_READY_TO_SEND);
+    CHK(pTlsSession->state != TLS_SESSION_STATE_CLOSED, STATUS_SOCKET_CONNECTION_CLOSED_ALREADY);
+
+    pReadBuffer = pTlsSession->pReadBuffer;
+    CHK_STATUS(ioBufferWrite(pReadBuffer, pData, *pDataLen));
+
+    // read application data
+    while (iterate && pReadBuffer->off < pReadBuffer->len && bufferLen > 0) {
+        sslRet = mbedtls_ssl_read(&pTlsSession->sslCtx, pData + readBytes, bufferLen);
+        if (sslRet > 0) {
+            readBytes += sslRet;
+            bufferLen -= sslRet;
+        } else if (sslRet == 0 || sslRet == MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY) {
+            // if sslRet is 0, the connection is closed already.
+            // if sslRet is MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY, the client notified us that the connection is going to be closed.
+            // In either case, we'll make sure that the state will change to CLOSED. If it's already closed, it'll be just a noop.
+            DLOGD("Detected TLS close_notify alert");
+            CHK_STATUS(tlsSessionShutdown(pTlsSession));
+            iterate = FALSE;
+        } else if (sslRet == MBEDTLS_ERR_SSL_WANT_READ || sslRet == MBEDTLS_ERR_SSL_WANT_WRITE) {
+            iterate = FALSE;
+        } else {
+            LOG_MBEDTLS_ERROR("mbedtls_ssl_read", sslRet);
+            readBytes = 0;
+            retStatus = STATUS_INTERNAL_ERROR;
+            iterate = FALSE;
+        }
+    }
+
+    if (pTlsSession->sslCtx.state == MBEDTLS_SSL_HANDSHAKE_OVER) {
+        tlsSessionChangeState(pTlsSession, TLS_SESSION_STATE_CONNECTED);
+    }
+
+CleanUp:
+    if (pDataLen != NULL) {
+        *pDataLen = readBytes;
+    }
+
+    // CHK_LOG_ERR might be too verbose
+    if (STATUS_FAILED(retStatus)) {
+        DLOGD("Warning: reading socket data failed with 0x%08x", retStatus);
+    }
+    
+    LEAVES();
+    return retStatus;
+}
+
+STATUS tlsSessionPutApplicationData(PTlsSession pTlsSession, PBYTE pData, UINT32 dataLen)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    UINT32 writtenBytes = 0;
+    BOOL iterate = TRUE;
```

Comment:
Hmm, I don't see why not since it's really just a flag marking if it should iterate, nothing else specific. I don't mind to rename it if you can come up with a better name.

Reply:
After looking a little more closely, I think it is ok.